core/domain/commit_commands_domain.py

```diff
@@ -0,0 +1,317 @@
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Domain object for commit commands."""
+
+from core.domain import collection_domain
+from core.domain import config_domain
+from core.domain import exp_domain
+from core.domain import story_domain
+import utils
+
+
+def validate_dict_keys(actual_keys, required_keys, optional_keys):
+    """Validates that actual keys in a dict contain all the required keys
+    and some/all of the optional keys.
+
+    Args:
+        actual_keys: list(str). The list of keys in parameters of a command.
+        required_keys: list(str). The list of keys which are required for
+            a command.
+        optional_keys: list(str). The list of keys which are optional for
+            a command.
+
+    Raises:
+        Exception. If the actual keys of parameters don't have all the
+            required keys or there are additional keys apart from the
+            required and optional keys.
+    """
+
+    missing_keys = [key for key in required_keys if key not in actual_keys]
+
+    extra_keys = [key for key in actual_keys if key not in (
+        required_keys + optional_keys)]
+
+    error_msg = ''
+    if missing_keys:
+        error_msg = error_msg + (
+            'The following required keys are missing: %s' % (
+                (',').join(missing_keys)))
+        if extra_keys:
+            error_msg = error_msg + ', '
+
+    if extra_keys:
+        error_msg = error_msg + 'The following extra keys are present: %s' % (
+            (',').join(extra_keys))
+
+    if missing_keys + extra_keys:
+        raise utils.ValidationError(error_msg)
+
+
+class BaseCommitCmd(object):
+    """Domain object for commit commands of a storage model.
+
+    Attributes:
+        name: str. The command name.
+        parameters: dict. The command parameters.
+    """
+
+    def __init__(self, name, parameters):
+        """Constructs a CommitCmd domain object.
+
+        Args:
+            name: str. The command name.
+            parameters: dict. The command parameters.
+        """
+        self.name = name
+        self.parameters = parameters
+        self.command_list = []
+
+    def validate(self):
+        """Checks that all parameters of a command are valid for the given
+        command.
+        """
+
+        required_keys = []
+        optional_keys = []
+        for cmd in self.command_list:
+            if cmd['name'] == self.name:
+                required_keys = cmd['required_keys']
+                optional_keys = cmd['optional_keys']
+                break
+        validate_dict_keys(self.parameters.keys(), required_keys, optional_keys)
+
+
+class ExplorationCommitCmd(BaseCommitCmd):
```

Comment:
The commands are defined in `exp_domain.py` and I used them in commit_commands_domain. Is the reverse better here? Defining the commands in the commands domain and then using them after validation elsewhere?

Reply:
Oh wait, hang on, maybe I am confused.

exp_domain.ExplorationChange seems to be describing a "commit cmd". So, what is this domain class for (how does it differ from ExplorationChange in terms of the object it represents)?

Reply:
There is no validation in `exp_domain` and it is not the class which describes the commands. The commands are defined in the file and then the arguments of the commands are used in ExplorationChange class. [Here](https://github.com/oppia/oppia/blob/develop/core/domain/exp_domain.py#L65) are the commands and [here](https://github.com/oppia/oppia/blob/develop/core/domain/exp_domain.py#L165) is the usage.

Reply:
Each command is a string. Why do you need a class to describe it?

It seems that what you are really describing here in ExplorationCommitCmd is an ExplorationChange (i.e. an exploration change dict) but that already exists as ExplorationChange right?

Reply:
Yes, but we do not have proper validation there. So, should we remove the commit commands and add a validation function to the Change class itself? Basically, we want to validate that old instances of commit commands in CommitLog models follow a valid schema. So, we need a validation function to ensure that all commit commands have a valid schema. 

The init in ExplorationChange will raise an error if something is missing since we are extracting property from dict but there is no check for extra args. So, a separate validation function can be added. What do you suggest?

Reply:
Yep, this sounds like the case where we should be adding the .validate() in the domain object itself (I made a more general comment about that) rather than defining a new class altogether. You could have that domain object subclass a base object which contains most of the validation logic, if you find yourself repeating lots of stuff across the different types of commit objects.

Reply:
Done.